returnn/extract_prior.py

```diff
@@ -58,3 +71,333 @@ def run(self):
+        import matplotlib
+
+        matplotlib.use("Agg")
+        import matplotlib.pyplot as plt
+
+        with open(self.out_prior_txt_file.get_path(), "rt") as f:
+            merged_scores = np.loadtxt(f, delimiter=" ")
+
+        xdata = range(len(merged_scores))
+        plt.semilogy(xdata, np.exp(merged_scores))
+        plt.xlabel("emission idx")
+        plt.ylabel("prior")
+        plt.grid(True)
+        plt.savefig(self.out_prior_png_file.get_path())
+
+    @classmethod
+    def create_returnn_config(
+        cls,
+        model_checkpoint,
+        returnn_config,
+        prior_data,
+        log_verbosity,
+        device,
+        **kwargs,
+    ):
+        """
+        Creates compute_prior RETURNN config
+        :param Checkpoint model_checkpoint:  TF model checkpoint. see `ReturnnTrainingJob`.
+        :param ReturnnConfig returnn_config: object representing RETURNN config
+        :param dict[str]|None prior_data: dataset used to compute prior (None = use one train epoch)
+        :param int log_verbosity: RETURNN log verbosity
+        :param str device: RETURNN device, cpu or gpu
+        :rtype: ReturnnConfig
+        """
+        assert device in ["gpu", "cpu"]
+        original_config = returnn_config.config
+        assert "network" in original_config
+
+        config = {
+            "load": model_checkpoint.ckpt_path,
+        }
+
+        config.update(copy.deepcopy(original_config))  # update with the original config
+
+        # override always
+        config["task"] = "compute_priors"
+
+        if prior_data is not None:
+            config["train"] = {
+                **original_config["train"].copy(),
+                **prior_data,
+            }
+
+        post_config = {
+            "device": device,
+            "log": ["./returnn.log"],
+            "log_verbosity": log_verbosity,
+        }
+
+        post_config.update(copy.deepcopy(returnn_config.post_config))
+
+        res = copy.deepcopy(returnn_config)
+        res.config = config
+        res.post_config = post_config
+        res.check_consistency()
+
+        return res
+
+    @classmethod
+    def hash(cls, kwargs):
+        d = {
+            "returnn_config": ReturnnComputePriorJob.create_returnn_config(**kwargs),
+            "returnn_python_exe": kwargs["returnn_python_exe"],
+            "returnn_root": kwargs["returnn_root"],
+        }
+        return super().hash(d)
+
+
+class ReturnnRasrComputePriorJob(ReturnnComputePriorJob, ReturnnRasrTrainingJob):
+    def __init__(
+        self,
+        train_crp,
+        dev_crp,
+        feature_flow,
+        model_checkpoint,
+        returnn_config,
+        prior_data=None,
+        alignment=None,
+        *,  # args below are keyword only
+        log_verbosity=3,
+        device="gpu",
+        time_rqmt=4,
+        mem_rqmt=4,
+        cpu_rqmt=1,
+        returnn_python_exe=None,
+        returnn_root=None,
+        # these are new parameters
+        num_classes=None,
```

Comment:
Are theses parameters all relevant for the PriorJob, or is it fine to leave them with their default values from the ReturnnRasrTrainingJob?

Reply:
Some of these parameters we afaik never set (e.g. `disregarded_classes`, `buffer_size`), but they are settable in the train job and I'd much rather not have them inconsistent.

Others (I mainly think of `partition_epochs`) make sense to set explicitly different than in the corresponding train Job

Reply:
If you are looking for consistency with the train job my suggestion would be to add **kwargs to the constructor and pass everything in kwargs to the train job. This way you also get forward compatibility.

Reply:
from https://github.com/rwth-i6/i6_core/blob/cef8a57cf0983b810c85c8f5507cc05a34ba4d19/returnn/rasr_training.py#L32 I thought we would get problems with Sisyphus when using the **kwargs here. Or is it ok in this case?

Reply:
Good question, I don't remember why passing **kwargs is/(was?) a problem at that point.